sql/core/src/main/scala/org/apache/spark/sql/ArrowConverters.scala

```diff
@@ -0,0 +1,432 @@
+  }
+
+  override def close(): Unit = {
+    byteArray = null
+  }
+
+  override def read(dst: ByteBuffer): Int = {
+    val remainingBuf = byteArray.length - _position
+    val length = Math.min(dst.remaining(), remainingBuf).toInt
+    dst.put(byteArray, _position.toInt, length)
+    _position += length
+    length
+  }
+
+  override def position(): Long = _position
+
+  override def position(newPosition: Long): SeekableByteChannel = {
+    _position = newPosition.toLong
+    this
+  }
+
+  override def size: Long = {
+    byteArray.length.toLong
+  }
+
+  override def write(src: ByteBuffer): Int = {
+    throw new UnsupportedOperationException("Read Only")
+  }
+
+  override def truncate(size: Long): SeekableByteChannel = {
+    throw new UnsupportedOperationException("Read Only")
+  }
+}
+
+/**
+ * Intermediate data structure returned from Arrow conversions
+ */
+private[sql] abstract class ArrowPayload extends Iterator[ArrowRecordBatch]
+
+/**
+ * Build a payload from existing ArrowRecordBatches
+ */
+private[sql] class ArrowStaticPayload(batches: ArrowRecordBatch*) extends ArrowPayload {
+  private val iter = batches.iterator
+  override def next(): ArrowRecordBatch = iter.next()
+  override def hasNext: Boolean = iter.hasNext
+}
+
+/**
+ * Class that wraps an Arrow RootAllocator used in conversion
+ */
+private[sql] class ArrowConverters {
+  private val _allocator = new RootAllocator(Long.MaxValue)
+
+  private[sql] def allocator: RootAllocator = _allocator
+
+  /**
+   * Iterate over the rows and convert to an ArrowPayload, using RootAllocator from this class
+   */
+  def interalRowIterToPayload(rowIter: Iterator[InternalRow], schema: StructType): ArrowPayload = {
+    val batch = ArrowConverters.internalRowIterToArrowBatch(rowIter, schema, _allocator)
+    new ArrowStaticPayload(batch)
+  }
+
+  /**
+   * Read an Array of Arrow Record batches as byte Arrays into an ArrowPayload, using
+   * RootAllocator from this class
+   */
+  def readPayloadByteArrays(payloadByteArrays: Array[Array[Byte]]): ArrowPayload = {
+    val batches = scala.collection.mutable.ArrayBuffer.empty[ArrowRecordBatch]
+    var i = 0
+    while (i < payloadByteArrays.length) {
+      val payloadBytes = payloadByteArrays(i)
+      val in = new ByteArrayReadableSeekableByteChannel(payloadBytes)
+      val reader = new ArrowReader(in, _allocator)
+      val footer = reader.readFooter()
+      val batchBlocks = footer.getRecordBatches.asScala.toArray
+      batchBlocks.foreach(block => batches += reader.readRecordBatch(block))
+      i += 1
+    }
+    new ArrowStaticPayload(batches: _*)
+  }
+
+  /**
+   * Call when done using this converter, will close RootAllocator so any ArrowBuffers should be
+   * closed first
+   */
+  def close(): Unit = {
+    _allocator.close()
+  }
+}
+
+private[sql] object ArrowConverters {
+
+  /**
+   * Map a Spark Dataset type to ArrowType.
+   */
+  private[sql] def sparkTypeToArrowType(dataType: DataType): ArrowType = {
```

Comment:
My latest update needed to use this outside of the object, but I will change it to `private[arrow]`